### ISSUE TITLE
[Question] HtmlParser called twice by the compiler

### ISSUE THREAD

**Issue by vicb** [angular/angular#4877]
The `HtmlParser` is called twice during compilation (by `TemplateParser` and `TemplateCompiler`).

Would it be possible to call it only once and re-use the AST ? 

I understand that this is not in the critical path and a potential refactoring should be given a low priority.

**Comment by tbosch** [comment]
Yes, we could, to improve the "on-the-fly" compile time.

Some restrictions for doing this:
We need to be able to serialize the CompileMetadata into a string and
deserialize it from there.

And we also should keep the interface of `TemplateParser` only take in a
string, so that someone could replace the TemplateParser altogether with an
own implementation that does not use the HtmlParser but does everything in
one go (e.g. transform JSX into a TemplateAST).

**Comment by mhevery** [comment]
Obsolete
